scylla/src/network/connection.rs

```diff
@@ -933,6 +933,30 @@ impl Connection {
         }
     }
 
+    fn get_result_id<'metadata>(
+        cached_metadata: &'metadata Option<Arc<ResultMetadata<'static>>>,
+        has_metadata_extension: bool,
+    ) -> Option<&'metadata [u8]> {
+        // The condition is basically equivalent to `has_metadata_extension`, but to get the
+        // metadata ouf of Option we need to do the match anyway. Why not just `if let`?
+        // Because if the extension is enabled, then we have to send the id.
+        // If the logic above is changed, and we have the extension, but no cached metadata,
+        // it would result in protocol error, so let's assert.
+        assert!(!has_metadata_extension || cached_metadata.is_some()); // has_metadata_extension -> cached_metadata.is_some()
+        match (cached_metadata.as_ref(), has_metadata_extension) {
+            // In the first branch we know that metadata extension is enabled, and we have extracted cached metadata.
+            // Can we `unwrap()` the id? I think not, because I see one scenario where it can be None.
+            // Statement may have been prepared on connection that does not have the extension (think:
+            // cluster during upgrade, with mixed versions), and then used on this connection.
+            // What to do in this case? We have to send some id, because it is not optional in the protocol,
+            // so let's send empty id.
+            (Some(metadata), true) => Some(metadata.id().unwrap_or(&[])),
+            (Some(_), false) => None,
+            (None, true) => unreachable!("metadata extension enabled, but no cached metadata."),
+            (None, false) => None,
```

Comment:
:pick: Let's merge both `false` cases to one, to highlight that absence of the extension makes things simple.